Added support for "`pragma protect ..." model encryption (#2886)
verilator/verilator: docs/guide/warnings.rst

   be declared before being used.

> .. option:: PRAGMAERR
> 
>    Any error that is related to pragmas, but not specifically to '`pragma protect'.
>    For example, an empty `pragma line.

Reviewer: ```suggestion
.. option:: BADPRAGMA

   Error that a pragma is badly formed, but not specifically to '`pragma protect'.
   For example, an empty `pragma line.
   
```

(And re-sort).  Generally don't want ERR in names as they can be changed into warnings a warning with ERR in the name would be confusing.

   a var/reg must be used as the target of procedural assignments.

> .. option:: PROTECTED
> 
>    PROTECTED indicates that a '`pragma protected' section was encountered, which
>    will be skipped by Verilator and partly checked for correctness, but is not
>    currently supported.
> 
> .. option:: PROTECTERR
> 
>    PROTECTERR indicates that errors were detected within a '`pragma protected'
>    section, such as incorrect BASE64 encoding, etc.
> 
> 

Reviewer: ```suggestion
.. option:: PROTECTED

   Warning that a '`pragma protected' section was encountered. The code
   inside the protected region will be partly checked for correctness, but is
   otherwise ignored.

   Suppressing the warning may make Verilator differ from a simulator that
   accepts the protected code.

.. option:: BADPROTECT

   Error that wasdetected within a '`pragma protected' section.
   such as incorrect BASE64 encoding, etc.

```

Alternatively IMO it would be fine to combine BADPROTECT into BADPRAGMA.

Author: I'll merge first, then merge fixes.

Author: I've given this a 2nd thought, and I believe keeping these separate has its advantage:
BADPROTECT is very specific - you were using a pragma that is properly specified in the spec in an incorrect way.
BADPRAGMA may trigger at the user's discretion to detect unknown pragmas although these could be safely ignored if desired, so you might want to turn this off while leaving BADPROTECT on.

Another option is to rename these to:
BADSTDPRAGMA - should be raised for any pragma that is clearly defined in the spec (there are others besides protect)
UNKNOWNPRAGMA (or something shorter) - should be raised as warning when an unknown pragma is encountered.
Accordingto IEEEE1800-2017, section 22.11, it should be safe to ignore those:
>Unless otherwise specified, pragma directives for
pragma_names that are not recognized by an implementation shall have no effect on interpretation of the
SystemVerilog source text.

Reviewer: I suggest we shouldn't warn on unknown pragma, given IEEE says not to.  Then we'd just have BADSTDPRAGMA.

Author: The fact that it shouldn't have any effect on interpretation doesn't mean you might want to know about such pragmas (even if you mark them as warnings or even Info, if you have such a distinction). Personally I would prefer to have more knobs even if some of them are even turned off by default.
However, as a maintainer that's your decision and I respect that.
I think in this case I'll use BADSTDPRAGMA  also for empty pragmas.
Just tested this with one of the Big-3 simulators (a 2019.x release), and an empty `pragma line was met with an internal exception, which is also a way to report an error :-)

Reviewer: >I think in this case I'll use BADSTDPRAGMA also for empty pragmas.
Sounds good, thanks.
